### PR TITLE
Add unit test framework with 98% line coverage

### LINKED ISSUE #52: Add unit test framework and tests with code coverage
## Summary

Introduce a unit test framework and initial test suite for metamod-p, with code coverage reporting. Follow the testing approach used in [jk_botti](https://github.com/Bots-United/jk_botti).

## Test framework

Select a lightweight C/C++ test framework suitable for the project. Key criteria:
- Easy to integrate with the existing Makefile build system
- Low dependency footprint
- Can be built as 32-bit (i686) to match the project target

## Makefile targets

- `make test` — build and run the test suite (`CXX="g++ -m32"`)
- `make valgrind` — run tests under valgrind to catch memory errors
- `make coverage` — build with gcov instrumentation, run tests, generate lcov report

## CI integration

Add test, valgrind, and coverage jobs to `ci.yml` (see #51):

- **test job**: run `make test` and `make valgrind`, install `valgrind` and `libc6-dbg:i386`
- **coverage job**: run `make coverage`, generate a coverage summary in the GitHub job summary (lines/functions/branches), upload the HTML coverage report as an artifact

## Initial test targets

Priority areas for unit tests, based on where bugs have been found historically:

- **String utilities** (`support_meta.h`): `STRNCPY`, `strmatch`, `strnmatch`, `safe_snprintf`, `safe_vsnprintf` — self-contained and easy to test
- **Config parsing** (`conf_meta.cpp`): `MConfig::load`, `MConfig::set` — config reload, value parsing, edge cases
- **Game lookup** (`game_support.cpp`): `lookup_game` — matching logic, duplicate name handling, fallback behavior
- **Plugin list operations** (`mlist.cpp`, `mplugin.cpp`): load, find, status transitions
- **Message registration** (`mreg.cpp`): `MRegMsgList::add`, `find`, duplicate handling, null name cases
- **Engine info detection** (`engineinfo.cpp`): `initialise`, `phdr_dladdr`, `phdr_r_debug` return value handling

## Considerations

- Many metamod functions depend on engine globals and DLL handles that need mocking/stubbing. Start with the self-contained utility functions and expand from there.
- Tests should build and run as a host-native binary (not requiring an actual HL engine to run).
- Keep test infrastructure minimal.

### PR DESCRIPTION
**Pull request by jkivilin**
## Summary

- Adds `metamod/tests/` with a self-contained test framework (no external test library deps): engine mock, Win32 API stubs, and custom assertion macros
- 35 test binaries covering all 29 metamod source files with real tests — 98.0% line coverage (4507/4597), 99.4% function coverage (617/621)
- CI jobs for test, valgrind (leak/error checking), and lcov coverage reporting

## Details

- Engine mock (`engine_mock.cpp/h`) stubs HL engine globals and provides capture buffers for AlertMessage/ServerPrint output, settable CMD_ARGV/ARGC, localinfo, cvars, and edict index control
- `test_common.h` provides ASSERT macros, temp file helpers, and `copy_test_plugin()` for CWD-independent test asset paths
- Tests for platform-specific code: dlsym hook patching (`osdep_linkent_linux`), ELF gamedll detection (`osdep_detect_gamedll_linux`), Win32 stubs compiled on Linux
- `#include "../source.cpp"` pattern used where needed to test static functions directly
- Mocked DLOPEN/DLSYM/DLCLOSE/DLERROR via preprocessor redirection for testing plugin lifecycle and gamedll loading without real shared libraries
- Valgrind integration: `RUNNING_ON_VALGRIND` skips tests incompatible with instrumentation (code patching)
- Malloc failure testing via `--wrap=calloc,--wrap=free` linker flags (`test_mreg_malloc`)

Closes #52.

### CODE REVIEW

File: metamod/tests/Makefile [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+test_commands_meta: test_commands_meta.o $(EXTENDED_OBJS) commands_meta.o vdate.o
+	${CXX} $(COVERAGE_FLAGS) -o $@ $^ -ldl
+
+test_dllapi: test_dllapi.o $(EXTENDED_OBJS) dllapi.o
+	${CXX} $(COVERAGE_FLAGS) -o $@ $^ -ldl
+
+test_engine_api: test_engine_api.o $(EXTENDED_OBJS) engine_api.o
+	${CXX} $(COVERAGE_FLAGS) -o $@ $^ -ldl
+
+test_h_export: test_h_export.o $(COMMON_OBJS) meta_eiface.o
```
[the remark this answers is not in the document]
Comment (jkivilin) [reply metamod/tests/Makefile]: By design — the goal of the test framework is to get all source files into the test build for coverage, not standalone compile testing. h_export.cpp is compiled via `#include` in test_h_export.cpp and is included in coverage.

File: metamod/tests/engine_mock.cpp [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+
+// ============================================================
+// mock_reset - reset all mock state between tests
+// ============================================================
+
+void mock_reset(void)
+{
+	mock_globalvars = globalvars_t();
+	mock_globalvars.maxClients = 32;
+	gpGlobals = &mock_globalvars;
```
[the remark this answers is not in the document]
Comment (jkivilin) [reply metamod/tests/engine_mock.cpp:376]: Fixed — mock_reset() now resets all globals: metamod_handle, Config, Plugins, Reg* lists, PublicMetaGlobals, PrivateMetaGlobals, hooked function pointers, g_Players, and g_plugin_engfuncs.

File: metamod/tests/Makefile [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+    test_game_support test_game_autodetect test_engineinfo \
+    test_sdk_util test_api_hook test_api_info \
+    test_commands_meta test_dllapi test_engine_api \
+    test_h_export test_linkgame test_linkplug \
+    test_meta_eiface test_metamod test_mlist test_mplugin \
+    test_mutil test_reg_support test_studioapi \
+    test_osdep_p test_osdep_linkent_linux test_osdep_detect_gamedll_linux
+
+all: $(ALL_TESTS)
+
```
[the remark this answers is not in the document]
Comment (jkivilin) [reply metamod/tests/Makefile:198]: Fixed — added root-level `make test`, `make valgrind`, and `make coverage` targets that delegate to `metamod/tests/`. The root `Makefile` now also cleans test artifacts on `make clean`.

File: metamod/tests/Makefile [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+	${CXX} $(COVERAGE_FLAGS) -o $@ $^ -ldl
+
+test_engineinfo: test_engineinfo.o $(COMMON_OBJS) engineinfo.o
+	${CXX} $(COVERAGE_FLAGS) -o $@ $^ -ldl
+
+test_sdk_util: test_sdk_util.o $(EXTENDED_OBJS)
+	${CXX} $(COVERAGE_FLAGS) -o $@ $^ -ldl
+
+test_vdate: test_vdate.o
+	${CXX} $(COVERAGE_FLAGS) -o $@ $^
```
[the remark this answers is not in the document]
Comment (jkivilin) [reply metamod/tests/Makefile:80]: By design — same as test_h_export. The goal is coverage inclusion, not standalone compile testing. vdate.cpp is compiled and covered via `#include` in the test file.

File: metamod/osdep_linkent_linux.cpp [hunk 5]
```diff
@@ -81,9 +84,9 @@ inline void construct_jmp_instruction(void *x, void *place, void* target)
 }
 
 //checks if pointer x points to jump forwarder
-inline bool is_code_trampoline_jmp_opcode(void *x) 
-{ 
-	return (((unsigned char *)x)[0] == 0xff || ((unsigned char *)x)[1] == 0x25); 
+inline bool is_code_trampoline_jmp_opcode(void *x)
+{
+	return (((unsigned char *)x)[0] == 0xff || ((unsigned char *)x)[1] == 0x25);
```
[the remark this answers is not in the document]
Comment (jkivilin) [reply metamod/osdep_linkent_linux.cpp:89]: This needs to be follow-up task. PR does not try to fix bugs in production code, but add missing unit-tests.

File: metamod/tests/test_osdep_linkent_linux.cpp [hunk 6]
```diff
@@ -0,0 +1,305 @@
...
+	PASS();
+	return 0;
+}
+
+static int test_trampoline_only_25(void)
+{
+	TEST("is_code_trampoline_jmp_opcode - 25 at byte[1] without FF (bug: || returns true)");
+	unsigned char code[] = { 0x00, 0x25, 0x00, 0x00, 0x00, 0x00 };
+	ASSERT_TRUE(is_code_trampoline_jmp_opcode(code) == true);
+	PASS();
```
[the remark this answers is not in the document]
Comment (jkivilin) [reply metamod/tests/test_osdep_linkent_linux.cpp:135]: This needs to be follow-up task. PR does not try to fix bugs in production code, but add missing unit-tests.

File: metamod/tests/test_mreg_malloc.cpp [hunk 7]
```diff
@@ -0,0 +1,330 @@
...
+	// BUG: production code leaks the calloc'd cvar_t when strdup fails
+	// (mreg.cpp line 337 returns without freeing icvar->data).
+	mock_alloc_fail_after(1);
+	MRegCvar *result = list.add("strdup_fail_cvar");
+	ASSERT_PTR_NULL(result);
+	ASSERT_TRUE(meta_errno == ME_NOMEM);
+	// Clean up the leaked cvar_t to satisfy valgrind
+	free(list.test_vlist()[list.test_endlist()].data);
+	list.test_vlist()[list.test_endlist()].data = NULL;
+
```
[the remark this answers is not in the document]
Comment (jkivilin) [reply metamod/tests/test_mreg_malloc.cpp:240]: We need to check if leak is intentional. HL engine might not provide interface for removing registered Cvars and thus we might not be able to free these (HL engine still use freed memory → corruption/segfault/etc).

File: metamod/tests/Makefile [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+	    --branch-coverage --ignore-errors inconsistent,deprecated
+	lcov --summary coverage.info --branch-coverage \
+	    --ignore-errors inconsistent,deprecated
+
+coverage-clean:
+	rm -f *.gcda *.gcno coverage.info
+	rm -rf coverage_report
+
+clean:
+	rm -f $(ALL_TESTS) *.o *.gcda *.gcno *.gcov coverage.info engine_test.so fake_gamedll.so
```
[the remark this answers is not in the document]
Comment (jkivilin) [reply metamod/tests/Makefile]: Fixed.

File: metamod/osdep_linkent_linux.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (jkivilin) [reply metamod/osdep_linkent_linux.cpp:89]: Tracked in #68.

File: metamod/tests/test_osdep_linkent_linux.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (jkivilin) [reply metamod/tests/test_osdep_linkent_linux.cpp:135]: Tracked in #68.
